hands-on-exercise/2-ignite-cli-adv/9-run-prod-docker.md

```diff
@@ -0,0 +1,755 @@
+---
+title: Simulate Production in Docker
+order: 12
+description: Run your checkers in a simulated production with Docker Compose
+tags:
+  - guided-coding
+  - cosmos-sdk
+  - dev-ops
+---
+
+# Simulate Production in Docker
+
+Before you launch yourself fully into production, it would be interesting to simulate a set of independent nodes on your own computer. This can be prepared and orchestrated with [Docker Compose](https://docs.docker.com/get-started/08_using_compose).
+
+## Target setup
+
+In order to mimic a real setup, find an objective on which to focus:
+
+* Three independent parties, Alice, Bob and Carol
+* Two independent validator nodes, run by Alice and Bob, that can only communicate with their own sentries and do not expose RPC end points.
+* One of the validator nodes, Alice's, uses Tendermint KMS.
+* The two sentry nodes, run by Alice and Bob, expose end points.
+* A regular node, run by Carol, that can communicate with only the sentries and exposes end points for use by clients.
+
+In terms of Docker concepts, this means:
+
+* Alice's validator and KMS are on the private user-defined network `net-alice-kms`.
+* Alice's validator and sentry are on the private user-defined network `net-alice`.
+* Bob's validator and sentry are on the private user-defined network `net-bob`.
+* There is a public user-defined network on which sentries and Carol's node run: `net-public`.
+
+## Docker elements
+
+Before looking at the specific Compose elements, you need to define what the _regular_ Docker elements are.
+
+### The node images
+
+The validators, sentries and regular nodes will run `checkersd`.
+
+In the [previous section](./8-run-prod.md), you built executables, in particular `checkersd-linux-amd64` and `checkersd-linux-arm64`. They should be enough for the purpose. Create a new `Dockerfile-ubuntu-prod` with:
+
+```Dockerfile [https://github.com/cosmos/b9-checkers-academy-draft/blob/run-prod/Dockerfile-ubuntu-prod]
+FROM --platform=linux ubuntu:22.04
+ARG BUILDARCH
+
+ENV LOCAL=/usr/local
+
+EXPOSE 1317 26656 26657
+
+COPY build/checkersd-linux-${BUILDARCH} ${LOCAL}/bin/checkersd
+
+ENTRYPOINT [ "checkersd" ]
+```
+
+And build the image with:
+
+```sh
+$ docker build -f Dockerfile-ubuntu-prod . -t checkersd_i
```

Comment:
I know you have said it L40, but it would be great to precise we need to run `make build-all` before (for those, like me, that just go to the commands :D)

Reply:
I needed (Fedora 35) to add `--build-arg BUILDARCH=amd64` for this to work.

Reply:
You computer is natively running Fedora 35 on an Intel architecture?

Reply:
Correct

Reply:
Can you install `buildx` and see if it magically works without `--build-arg`?
https://github.com/docker/buildx

Reply:
It did!